File: my-solutions/0004-median-of-two-sorted-arrays/solution.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A, B = nums1, nums2
        if len(A) > len(B):
            B, A = A, B

        l, r = 0, len(A) - 1
        size = len(nums1) + len(nums2)
        while(True):
            i = (l + r) // 2
            j = size // 2 - i - 2

            # print(l, r, i, j)

            Aleft = A[i] if i >= 0 else float('-inf')
            Aright =  A[i+1] if i + 1 < len(A) else float('inf')
            Bleft = B[j] if j >= 0 else float('-inf')
            Bright = B[j + 1] if j + 1 < len(B) else float('inf')

            
            # print(Aleft, Bleft, Aright, Bright)
            # print("")

            if Aleft <= Bright and Bleft <= Aright:
                if size % 2 == 0:
                    return (min(Aright, Bright) + max(Aleft, Bleft)) / 2
                else:
                    return min(Aright, Bright)

            elif Aleft > Bright:
                r = i - 1
            else:
                l = i + 1

            

File: my-solutions/0004-median-of-two-sorted-arrays/solution.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size(), n = nums2.size();
        if(m > n) return findMedianSortedArrays(nums2, nums1);        
        int l = 0, r = m;
        while(l <= r) {
            int x = (l + r) / 2;
            int y = (m + n + 1) / 2 - x;

            int l1 = (x == 0) ? INT_MIN : nums1[x - 1];
            int r1 = (x == m) ? INT_MAX : nums1[x];
            int l2 = (y == 0) ? INT_MIN : nums2[y - 1];
            int r2 = (y == n) ? INT_MAX : nums2[y];

            if(l1 <= r2 && l2 <= r1) {
                if((m + n)%2) return max(l1, l2);
                return double(max(l1, l2) + min(r1, r2)) / 2;
            }
            if(l1 > r2) r = x - 1;
            else l = x + 1;
        }

        return 0.0;
    }
};
